scripts/create_changelog.py

```diff
@@ -0,0 +1,119 @@
+        if START_TOKEN.lower() in line[: len(START_TOKEN) + 3].lower():
+            found = True
+    return None
+
+
+def _normalize(name: str) -> str:
+    return name.replace(" ", "").lower()
+
+
+def _is_external_commit(commit: git.Commit) -> bool:
+    # not super accurate at the moment, we'll probably need to actually ping the Github API
+    return bool(commit.co_authors) and any(
+        _normalize(str(a.name)) != _normalize(str(commit.author.name)) for a in commit.co_authors
+    )
+
+
+def _get_completed_sections(new_version: str) -> str:
+    completed = f"# Changelog {new_version}"
+    completed += (
+        f"\n## {new_version} (core) / {_get_libraries_version(new_version)} (libraries)\n\n"
+    )
+    sections = ["New", "Bugfixes", "Breaking Changes", "Deprecations", "Documentation", "Unsorted"]
+    for section in sections:
+        completed += f"### {section}\n\n"
+    return completed
+
+
+def _update_changelog(new_version: str, prev_version: str) -> None:
+    repo = git.Repo(REPO_ROOT)
+    documented = []
+    undocumented = defaultdict(list)
+
+    for commit in repo.iter_commits(rev=f"release-{prev_version}..release-{new_version}"):
+        if IGNORE_TOKEN in str(commit.message):
+            continue
+
+        title = str(commit.message).splitlines()[0]
+        # me avoiding regex -- titles are formatted as "Lorem ipsum ... (#12345)" so we can just search
+        # for the last octothorpe and chop off the closing paren
+        issue_number = title.split("#")[-1][:-1]
+        issue_link = f"[#{issue_number}]({GITHUB_URL}/pull/{issue_number})"
+
+        changelog_msg = _get_changelog_msg(commit)
+
+        if changelog_msg is None:
+            # rough filter for detecting external PRs
+            if _is_external_commit(commit):
+                responsible_party = commit.co_authors[-1].name
+                thanks = f" (Thanks @{commit.author.name}[TODO]!)"
+            else:
+                responsible_party = commit.author.name
+                thanks = ""
+
+            # group undocumented commits by the person responsible for documenting them
+            undocumented[responsible_party].append(f"{issue_link} {title}{thanks}")
+        else:
+            documented.append(f"* {issue_link} {changelog_msg}")
+
+    uncompleted_text = "# Undocumented Changes"
+    for responsible_party, commits in sorted(undocumented.items()):
+        uncompleted_text += f"\n\n- [ ] {responsible_party}"
+        for commit in commits:
+            uncompleted_text += f"\n\t- [ ] {commit}"
+
+    completed_text = _get_completed_sections(new_version)
+    for commit in documented:
+        completed_text += f"\n{documented}"
+
+    sys.stdout.write(completed_text + uncompleted_text)
+
+
+@click.command()
+@click.argument("new_version", type=str, required=True)
+@click.argument("prev_version", type=str, required=False)
+def update_changelog(new_version: str, prev_version: Optional[str] = None) -> None:
```

Comment:
`generate_changelog` perhaps? `update_changelog` makes it sound like it mutates some state, but it's really just outputting text, right?

Reply:
updated